src/import/FSHImporter.ts

```diff
@@ -1384,8 +1422,110 @@ export class FSHImporter extends FSHVisitor {
+              );
+              return insertRule;
+            } else {
+              logger.error(
+                `Failed to parse RuleSet ${
+                  insertRule.ruleSet
+                } with provided parameters (${insertRule.params.join(', ')})`,
+                insertRule.sourceInfo
+              );
+            }
+          }
+        } else {
+          logger.error(
+            `Incorrect number of parameters applied to RuleSet ${insertRule.ruleSet}`,
+            insertRule.sourceInfo
+          );
+        }
+      } else {
+        logger.error(
+          `Could not find RuleSet with parameters named ${insertRule.ruleSet}`,
+          insertRule.sourceInfo
+        );
+      }
+    } else {
+      return insertRule;
+    }
+  }
+
+  private parseRulesetReference(reference: string): [string, string] {
+    const paramListStart = reference.indexOf('(');
+    if (paramListStart === -1) {
+      return [reference.trim(), null];
+    } else {
+      return [reference.slice(0, paramListStart).trim(), reference.slice(paramListStart)];
+    }
+  }
+
+  private parseInsertRuleParams(ruleText: string): string[] {
+    // first, trim parentheses
+    const ruleNoParens = ruleText.slice(1, ruleText.length - 1);
+    // since backslash is the escape character, deal with literal backslash first
+    const splitBackslash = ruleNoParens.split(/\\\\/g);
+    // then, split the parameters apart with unescaped commas
+    const splitComma = splitBackslash.map(substrBackslash => {
+      return substrBackslash.split(/(?<!\\),/g).map(substrComma => {
+        // then, make all the replacements: closing parenthesis and comma
+        return substrComma.replace(/\\\)/g, ')').replace(/\\,/g, ',');
+      });
+    });
+    const paramList: string[] = [];
+    // if splitComma has more than one list, that means we split on literal backslash
+    // so to rejoin all the strings, the last string joins the first string in the next sublist
+    splitComma.forEach((list, index) => {
+      list.forEach((paramPart, subIndex) => {
+        if (index > 0 && subIndex === 0) {
+          // join with \\ on the last param
+          paramList.push(`${paramList.pop()}\\\\${paramPart}`);
+        } else {
+          // push a new param
+          paramList.push(paramPart);
+        }
+      });
+    });
+    // trim whitespace from each parameter, since it may be formatted for readability
+    return paramList.map(param => param.trim());
+  }
+
+  private parseGeneratedRuleSet(input: string, name: string) {
+    // define a temporary document that will contain this RuleSet
+    const tempDocument = new FSHDocument(this.currentFile);
+    // save the currentDoc so it can be restored after parsing this RuleSet
+    const parentDocument = this.currentDoc;
+    this.currentDoc = tempDocument;
+    const subContext = this.parseDoc(input, this.currentFile);
+    this.visitDoc(subContext);
+    // restore parentDocument
+    this.currentDoc = parentDocument;
```

Comment:
If `this.parseDoc` or `this.visitDoc` throws, then `this.currentDoc` will never get set back.  I'm guessing that if they throw maybe we bomb out of everything and it doesn't matter anyway, but... in case we _do_ gracefully recover (now or in the future), it might be safer to wrap lines 1517-1518 in `try` and put line 1520 in a `finally` -- just to guarantee we set it back no matter what happens.